### ISSUE TITLE
Provide ability to read in geny files

### ISSUE THREAD

**Issue by namurphy** [sunpy/sunpy#5185]
Some of the data files in SSWIDL are in `.geny` format.  For example, some files relevant to [xrtpy](https://github.com/HinodeXRT/xrtpy) are in the [directory containing response information for different channels for Hinode/XRT](https://hesperia.gsfc.nasa.gov/ssw/hinode/xrt/idl/response/channels/).  Following the question by @joyvelasquez in #1927 , it'd be helpful to add the capability to read in `geny` files too.

**Comment by namurphy** [comment]
We did try using `scipy.io.special.read_genx` but got this error:
```Python
>>> f = read_genx('/home/namurphy/Downloads/xrt_channels_v0016.geny')
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-9-8dbd5ea77ad0> in <module>
----> 1 f = read_genx('/home/namurphy/Downloads/xrt_channels_v0016.geny')

~/.miniconda/lib/python3.9/site-packages/sunpy/io/special/genx.py in read_genx(filename)
    174     # the number of upper level strs
    175     # This is used somehow
--> 176     mainsize = arr_size[2]  # NOQA
    177 
    178     skeleton = read_struct_skeleton(xdrdata)

IndexError: list index out of range
```

@joyvelasquez sent me the contents of `xrt_channels_v0016.geny` from within IDL:
```
IDL> help,channels, /str
** Structure XRT_CHN_CONFIG_V0016, 22 tags, length=323600, data length=323556:
   TYPE            STRING    'XRT_CHN_CONFIG'
   CHN_STR_VERSION STRING    'XRT_CHN_CONFIG_V0016'
   CHN_STR_DESCR   STRING    'Specifies an XRT instrument configuration, without the transmissi'...
   NAME            STRING    'Al-mesh'
   LONG_NAME       STRING    'Al-mesh XRT channel'
   OBSERVATORY     STRING    'Hinode'
   INSTRUMENT      STRING    'XRT'
   GEOM            STRUCT    -> XRT_GEOMETRY_V0001 Array[1]
   EN_FILTER       STRUCT    -> XRT_FILTER_V0001 Array[1]
   MIRROR1         STRUCT    -> XRT_MIRROR_V0001 Array[1]
   MIRROR2         STRUCT    -> XRT_MIRROR_V0001 Array[1]
   FP_FILTER1      STRUCT    -> XRT_FILTER_V0001 Array[1]
   FP_FILTER2      STRUCT    -> XRT_FILTER_V0001 Array[1]
   CONTAM          STRUCT    -> XRT_FILTER_V0001 Array[1]
   CCD             STRUCT    -> XRT_CCD_V0001 Array[1]
   WAVE            FLOAT     Array[5000]
   WAVE_UNITS      STRING    'Angstroms'
   TRANS           FLOAT     Array[5000]
   LENGTH          LONG              3993
   CONFIG_FILENAME STRING    'xrt_config_v0016'
   HISTORY         STRING    Array[3]                                    
   COMMENTS        STRING    Array[5]
```

We also tried using `scipy.io.readsav` and got a `np.recarray`, but this ended up being pretty difficult to work with.

**Comment by Cadair** [comment]
If `scipy.io.readsav` loads them then they are IDL sav files, and we don't need to implement a reader in SunPy (and probably can't as the scipy one got approval from the owners of IDL).  What about `readsav` was hard to work with?

**Comment by hayesla** [comment]
fwiw, I usually just convert the `np.recarray` to a dictionary when reading from a `sav` file
```python
data = readsav("xrt_channels_v0016.geny")["p0"]
data_dict = {k : data[k] for k in data.dtype.names}
```
I find this kind of easier as then the variables are accessed through a way easier for python users, i.e. `data_dict["NAME"]` rather than `data.NAME` etc.

**Comment by nabobalis** [comment]
I think based on Laura's comment, I don't think we need a new reader for this file type. 

Maybe an example or a page somewhere in our docs might be useful?

**Comment by Cadair** [comment]
> I find this kind of easier as then the variables are accessed through a way easier for python users, i.e. data_dict["NAME"] rather than data.NAME etc.

But you can access recarrays like this, and in fact do in your example? (I am confused :confused:)

**Comment by hayesla** [comment]
yes, that's true!  It's literally just for convenience for those not used to dealing with `recarray`s- easier to do things like `data_dict.keys()` or can be easily readable when printing. 

But yes I agree, it doesn't really make a difference ,I've just found it something that I generally do when dealing with these

**Comment by namurphy** [comment]
Thank you — we'll give @hayesla's example a go.  The main reason I brought up this issue is due to [this comment](https://github.com/sunpy/sunpy/issues/1927#issuecomment-730027115) in #927 and because SunPy does have `sunpy.io.special.genx.read_genx`.

Looking back, one reason we got confused is that `.genx` and `.geny` appear to be quite different...in contrast to what we interpreted from the extension.  It appears that the former corresponds to the External Data Representation Standard file format (XDR) whereas the latter is an IDL `.sav` file with the extension renamed.  I'm not sure why `.geny` was used instead of `.sav` though.

There's a chance that people in the future could get confused about this just like we did.  A note in the docstring of `read_genx` that it is not intended for `.geny` files could prevent future confusion, but then again having the discussion in this issue would probably be enough.  I'll try to remember to post again here about how we end up addressing this in xrtpy.  Thank you again!

**Comment by Cadair** [comment]
We could even get fancy in the genx reader and detect if the file isn't a XDR and then throw an error saying something like "this isn't genx, have you tried to load it with readsav"?

**Comment by namurphy** [comment]
In case there's anyone in the future who is trying to work with `.geny` files in Python and happens across this issue, the workaround that we decided on using was to convert the `.geny` file to a `.genx` file in IDL, and then use `read_genx` on the new `.genx` file.  According to @joyvelasquez, `read_genx` is putting the new `.genx` file into a very usable format.  This particular file has structures within arrays within structures (or something complicated like that), so the functionality from `read_genx` is turning out to be really helpful with that.  For simpler `.geny` files, `scipy.io.readsav` is still a very workable solution.  One cautionary note is that an apparent difference between `.genx` and `.geny` is that `.geny` files can handle objects and pointers in IDL but that `.genx` can't.  

Here's an example of using SSWIDL to convert a `.geny` to a `.genx` file (with thanks to Kathy Reeves).  
> ```
> IDL> ;; Read in geny file:
> IDL> restgenx, file='xrt_channels_v0016.geny', channels
> 
> IDL> ;; Save as genx file:
> IDL> savegen, file ='/path/to/file/xrt_channels_v0016.genx', channels
> ```
> (Note that functions that end in `x`, `readgenx`, `savegenx`, are, confusingly, for `.geny` files.  Functions without the `x`, `restgen` and `savegen`, are for `.genx` files).

And with that, we can now rest easy that this issue doesn't end up as a repeat of [this xkcd](https://xkcd.com/979/) in ∼2036.

**Comment by Cadair** [comment]
I am now going to repurpose this issue to make some improvements. The things I think we should do here are:

* Add a "See Also" section in the genx reader docstring about using `scipy.io.readsav` to load "geny" files.
* Detect in the genx loader if the file isn't an XDR file and issue a nicer error.
* Detect in the genx loader if the file is infact a IDL save file, and issue and even nicer error saying "you want to use scipy.io.readsav".

**Comment by sr-dash** [comment]
I'd like to work on this one. Following @Cadair 's list above. 
I have a local branch fixing this with
1. read_genx checks first four bytes for IDL sav file type signature and raises with a message pointing at scipy.io.readsav
2. "See also" entry for scipy.io.readsav to the read_genx docstring saying .geny files are also IDL save files.

**Comment by Cadair** [comment]
Hi @sr-dash that sounds great, please open a PR thanks!
